pkg/controller/external_terraform_plugin_framework.go

```diff
@@ -0,0 +1,642 @@
+		}
+		opTracker.SetFrameworkTFState(tfStateDynamicValue)
+	}
+
+	configuredProviderServer, err := c.configureProvider(ctx, ts)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not configure provider server")
+	}
+
+	return &terraformPluginFrameworkExternalClient{
+		ts:             ts,
+		config:         c.config,
+		logger:         logger,
+		metricRecorder: c.metricRecorder,
+		opTracker:      opTracker,
+		resource:       c.config.TerraformPluginFrameworkResource,
+		server:         configuredProviderServer,
+		params:         params,
+		resourceSchema: resourceSchema,
+	}, nil
+}
+
+func (c *TerraformPluginFrameworkConnector) getResourceSchema(ctx context.Context) (rschema.Schema, error) {
+	res := c.config.TerraformPluginFrameworkResource
+	schemaResp := &fwresource.SchemaResponse{}
+	res.Schema(ctx, fwresource.SchemaRequest{}, schemaResp)
+	if schemaResp.Diagnostics.HasError() {
+		return rschema.Schema{}, errors.Errorf("could not retrieve resource schema: %v", schemaResp.Diagnostics)
+	}
+
+	return schemaResp.Schema, nil
+}
+
+func (c *TerraformPluginFrameworkConnector) configureProvider(ctx context.Context, ts terraform.Setup) (tfprotov5.ProviderServer, error) {
+	var schemaResp fwprovider.SchemaResponse
+	ts.FrameworkProvider.Schema(ctx, fwprovider.SchemaRequest{}, &schemaResp)
+	if schemaResp.Diagnostics.HasError() {
+		var diagErrors []string
+		for _, tfdiag := range schemaResp.Diagnostics.Errors() {
+			diagErrors = append(diagErrors, fmt.Sprintf("%s: %s", tfdiag.Summary(), tfdiag.Detail()))
+		}
+		return nil, fmt.Errorf("cannot retrieve provider schema: %s", strings.Join(diagErrors, "\n"))
+	}
+	providerServer := providerserver.NewProtocol5(ts.FrameworkProvider)()
+
+	providerConfigDynamicVal, err := protov5DynamicValueFromMap(ts.Configuration, schemaResp.Schema.Type().TerraformType(ctx))
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot construct dynamic value for TF provider config")
+	}
+
+	configureProviderReq := &tfprotov5.ConfigureProviderRequest{
+		TerraformVersion: "crossTF000",
+		Config:           providerConfigDynamicVal,
+	}
+	providerResp, err := providerServer.ConfigureProvider(ctx, configureProviderReq)
+	if err != nil {
+		return nil, err
+	}
+	if fatalDiags := getFatalDiagnostics(providerResp.Diagnostics); fatalDiags != nil {
+		return nil, errors.Wrap(fatalDiags, "provider configure request failed")
+	}
+	return providerServer, nil
+}
+
+func (n *terraformPluginFrameworkExternalClient) getDiffPlan(ctx context.Context,
+	tfStateValue tftypes.Value) (*tfprotov5.DynamicValue, bool, error) {
+
+	valueTerraformType := n.resourceSchema.Type().TerraformType(ctx)
+
+	tfConfigDynamicVal, err := protov5DynamicValueFromMap(n.params, valueTerraformType)
+	if err != nil {
+		return &tfprotov5.DynamicValue{}, false, errors.Wrap(err, "cannot construct dynamic value for TF Config")
+	}
+
+	//
+	tfPlannedStateDynamicVal, err := protov5DynamicValueFromMap(n.params, valueTerraformType)
```

Comment:
This is defensively coded as they are passed to `PlanResourceChange()` method as pointers and the underlying implementation can potentially modify plan content and affect config content while doing this. So I opted for calculating it twice to make it safer and code looks simpler. DynamicValue does not offer any Copy function. It can be optimized though, by calculating the TF value only for config and calling its Copy() method, then creating two `DynamicValue`s.